RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc

```diff
@@ -651,16 +651,11 @@ bool DAClusterizerInZ_vect::purge(vertex_t& y, track_t& tks, double& rho0, const
     double sump = 0;
 
     double pmax = y.pk_ptr[k] / (y.pk_ptr[k] + rho0 * local_exp(-beta * dzCutOff_ * dzCutOff_));
-    for (unsigned int i = 0; i < nt; i++) {
-      //if (tks.tt[i]->track().pt() < 0.0) continue; // TEST count only pt>0.3 in nunique
-      if (tks.Z_sum_ptr[i] > 1.e-100) {
-        double p = y.pk_ptr[k] * local_exp(-beta * Eik(tks.z_ptr[i], y.z_ptr[k], tks.dz2_ptr[i])) / tks.Z_sum_ptr[i];
-        //double p = tks.pi_ptr[i] * y.pk_ptr[k] * local_exp(-beta * Eik(tks.z_ptr[i], y.z_ptr[k], tks.dz2_ptr[i])) / tks.Z_sum_ptr[i];
-        sump += p;
-        if ((p > uniquetrkweight_ * pmax) && (tks.pi_ptr[i] > 0)) {
-          nUnique++;
-        }
-      }
+    for (unsigned int i = 0; i < nt; ++i) {
+      auto ptr = y.pk_ptr[k] * local_exp(-beta * Eik(tks.z_ptr[i], y.z_ptr[k], tks.dz2_ptr[i]));
```

Comment:
There are four unique vectors involved in this calculations. Should I copy all of them into temporary arrays.
Eik calls std::pow with two arguments and multiplies with third
local_exp calls vdt::fast_math
so this becomes
```
auto ptr = y.pk_ptr[k] * vdt::fast_exp ( -beta * std::pow(tks.z_ptr[i], y.z_ptr[k]) * tks.dz2_ptr[i] )
```

Reply:
please verify the code of Eik again...

Reply:
https://github.com/cms-sw/cmssw/blob/217d7faafb922a4cc50b577acef6e03dd9ef8a32/RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZ_vect.cc#L215

Reply:
It's fixed.

Reply:
why copying? gcc is very bad with "restrict" ok. still a copy is expensive

Reply:
> change the std::vector in a local array

The vectors are member data. Can the be changed into a local array?

Reply:
No. they should be allocated on the stack

Reply:
why are member data?

Reply:
ok I see. the raw pointers are declred restrict 
https://github.com/cms-sw/cmssw/blob/217d7faafb922a4cc50b577acef6e03dd9ef8a32/RecoVertex/PrimaryVertexProducer/interface/DAClusterizerInZ_vect.h#L67
in my opinion gcc "optmize that out" before the vectorizer has any chance to see them.

Reply:
https://github.com/cms-sw/cmssw/blob/af0491716a5e551e527bc33c4e995e5870823467/RecoVertex/PrimaryVertexProducer/interface/DAClusterizerInZT_vect.h#L200

Reply:
> why are member data?
I assumed that y and tks were class references from the access pattern.
I see now that y and tks are struct references.